crypto/cms/cms_sd.c

```diff
@@ -547,8 +547,11 @@ EVP_MD_CTX *CMS_SignerInfo_get0_md_ctx(CMS_SignerInfo *si)
 
 STACK_OF(CMS_SignerInfo) *CMS_get0_SignerInfos(CMS_ContentInfo *cms)
 {
-    CMS_SignedData *sd = cms_get0_signed(cms);
+    CMS_SignedData *sd;
 
+    ERR_set_mark();
```

Comment:
Why is this needed here?

Reply:
This is explained in the commit message:
```
CMS_get0_SignerInfos(): Pervent spurious error on cms_get0_signed() failure
```

Reply:
Hmm, right. I did not look at the commits individually. There is also a typo in the commit message.

Did you get any insight why the spurious error appears? Would it be better to fix it at the place?

Reply:
I just re-discovered the stacktrace for it:
```
	cms_get0_signed() at cms_sd.c:30 0x7ffff7b83a98	
	CMS_get0_SignerInfos() at cms_sd.c:553 0x7ffff7b84fc8	
	ossl_cms_SignerInfos_set_cmsctx() at cms_sd.c:499 0x7ffff7b84e52	
	ossl_cms_resolve_libctx() at cms_lib.c:101 0x7ffff7b809a4	
	SMIME_read_CMS_ex() at cms_io.c:101 0x7ffff7b7f48d	
	load_content_info() at cms.c:242 0x4353e6	
	cms_main() at cms.c:896 0x436dce	
	do_cmd() at openssl.c:414 0x44dbb3	
	main() at openssl.c:295 0x44d7e2	
```
and found a more elegant solution: replace the error queue manipulation by
```
    if (OBJ_obj2nid(cms->contentType) != NID_pkcs7_signed)
        return NULL;
```

Reply:
> There is also a typo in the commit message.

Opos, fixed as well.

Reply:
I would actually do it differently. Fix it in ossl_cms_SignerInfos_set_cmsctx() - just put the set_mark/pop_to_mark calls there.

Reply:
My reasoning - the problem is because the set_cmsctx() calls CMS_get0_SignerInfos() regardless of the contentType, not because the CMS_get0_SignerInfos() should not report the error. Or, alternatively it might make sense to drop the ERR_raise completely, but I am not sure about that.

Reply:
Ok, if you prefer. Done.

Reply:
Just saw your reasoning. Sounds reasonable :)